image-loader/app/model/ImageUpload.scala

```diff
@@ -28,56 +75,73 @@ case object ImageUpload {
-        sourceDimensions <- sourceDimensionsFuture
-        thumbDimensions  <- thumbDimensionsFuture
-        fileMetadata     <- fileMetadataFuture
-        colourModel      <- colourModelFuture
-        fullFileMetadata  = fileMetadata.copy(colourModel = colourModel)
-
-        metadata      = ImageMetadataConverter.fromFileMetadata(fullFileMetadata)
-        cleanMetadata = metadataCleaners.clean(metadata)
-
-        sourceAsset = Asset.fromS3Object(s3Source, sourceDimensions)
-        thumbAsset  = Asset.fromS3Object(s3Thumb,  thumbDimensions)
-
-        baseImage      = createImage(uploadRequest, sourceAsset, thumbAsset, fullFileMetadata, cleanMetadata)
-        processedImage = SupplierProcessors.process(baseImage)
-
-        // FIXME: dirty hack to sync the originalUsageRights and originalMetadata as well
-        finalImage     = processedImage.copy(
-          originalMetadata    = processedImage.metadata,
-          originalUsageRights = processedImage.usageRights
-        )
-      }
-      yield {
-        ImageUpload(uploadRequest, finalImage)
+    fileMetadataFuture.flatMap(fileMetadata => {
+
+      // These futures are started outside the for-comprehension, otherwise they will not run in parallel
+      val sourceStoreFuture = storeSource(uploadRequest)
+      // FIXME: pass mimeType
+      val colourModelFuture = ImageOperations.identifyColourModel(uploadedFile, "image/jpeg")
+      val sourceDimensionsFuture = FileMetadataReader.dimensions(uploadedFile, uploadRequest.mimeType)
+
+      val thumbFuture = for {
+        fileMetadata <- fileMetadataFuture
+        colourModel <- colourModelFuture
+        iccColourSpace = FileMetadataHelper.normalisedIccColourSpace(fileMetadata)
+        thumb <- ImageOperations.createThumbnail(uploadedFile, thumbWidth, thumbQuality, tempDir, iccColourSpace, colourModel)
+      } yield thumb
+
+
+      val optimisedPng = OptimisedPng.build(uploadedFile, uploadRequest, fileMetadata)
+
+      bracket(thumbFuture)(_.delete) { thumb =>
+        // Run the operations in parallel
+        val thumbStoreFuture = storeThumbnail(uploadRequest, thumb)
+
+        val thumbDimensionsFuture = FileMetadataReader.dimensions(thumb, Some("image/jpeg"))
+
+        for {
+          s3Source <- sourceStoreFuture
+          s3Thumb <- thumbStoreFuture
+          s3PngOption <- optimisedPng.optimisedFileStoreFuture
+          sourceDimensions <- sourceDimensionsFuture
+          thumbDimensions <- thumbDimensionsFuture
+          fileMetadata <- fileMetadataFuture
+          colourModel <- colourModelFuture
+          fullFileMetadata = fileMetadata.copy(colourModel = colourModel)
+
+          metadata = ImageMetadataConverter.fromFileMetadata(fullFileMetadata)
+          cleanMetadata = metadataCleaners.clean(metadata)
+
+          sourceAsset = Asset.fromS3Object(s3Source, sourceDimensions)
+          thumbAsset = Asset.fromS3Object(s3Thumb, thumbDimensions)
+
+          pngAsset = if (optimisedPng.isPng24)
```

Comment:
Getting a pngAsset could be method but not a property I think because calculating it requires knowing the source dimensions and s3PngOption, futures that are resolved just a few lines above this (although I guess calculating the s3PngOption future could be moved to pngAsset). The method would still require the isPng24 check which is annoying.